KB-12170 | BE|DEV| Create new API for listing the Microcredentials content

tarentomaheshvakkund opened the pull request:
…ntent

1.  New API added to add the rules for  promotional content  - the table design is same is access access_setting_rules_v2 .
2. Delete API is a soft delete API.
3.  Added new promotional API which fetches the data as per the current my assigned courses and returns the response .
4.  Added this new cache manager class PromotionalContentRuleCacheMgr to handle the caching for the promotional content.
5. Added new conifg property specific for reading the content for promotional content.

tarentomaheshvakkund committed:
KB-12170 | BE|DEV| Create new API for listing the Microcredentials content

1.  New API added to add the rules for  promotional content  - the table design is same is access access_setting_rules_v2 .
2. Delete API is a soft delete API.
3.  Added new promotional API which fetches the data as per the current my assigned courses and returns the response .
4.  Added this new cache manager class PromotionalContentRuleCacheMgr to handle the caching for the promotional content.
5. Added new conifg property specific for reading the content for promotional content.

tarentomaheshvakkund committed:
Junit test cases added

karthik-tarento commented on src/main/java/com/igot/cb/cache/PromotionalContentRuleCacheMgr.java:
```diff
@@ -0,0 +1,214 @@
+
+    private final RedisCacheMgr redisCacheMgr;
+    private final CassandraOperation cassandraOperation;
+    private Map<String, CachedAccessSettingRule> cachedAccessSettingRules = new ConcurrentHashMap<>();
```
this needs to be done using Caffine Cache - refer the other classes.

tarentomaheshvakkund replied:
Implemented the caffine cache

karthik-tarento commented on src/main/java/com/igot/cb/cache/PromotionalContentRuleCacheMgr.java:
```diff
@@ -0,0 +1,214 @@
+                                Map.Entry::getKey,
+                                entry -> new CachedAccessSettingRule(entry.getValue())));
+            } else {
+                List<Map<String, Object>> accessSettingRuleMapList = cassandraOperation.getRecordsByProperties(
```
what are the other possibilities -- after few months if we have 1000s of records, this would become performance heavy read...

tarentomaheshvakkund replied:
Implemented the pagination for fetching records from the cassandra db .

karthik-tarento commented on src/main/java/com/igot/cb/service/PromotionalContentServiceImpl.java:
```diff
@@ -0,0 +1,349 @@
+        if (StringUtils.isEmpty(userId)) {
+            String errMsg = "Invalid or missing authentication token";
+            log.error(errMsg);
+            response.updateErrorDetails(errMsg, HttpStatus.UNAUTHORIZED);
```
this should be done inside accessTokenValidator -- if not, then update accordingly..

tarentomaheshvakkund replied:
fixed it

karthik-tarento commented on src/main/java/com/igot/cb/service/PromotionalContentServiceImpl.java:
```diff
@@ -0,0 +1,349 @@
+        if (MapUtils.isEmpty(userGroupDetails)) {
+            log.error("User group details are null or empty");
+            setFailedResponse(response, "User group details cannot be null or empty");
+            return response;
```
move this if inside validation method only.

tarentomaheshvakkund replied:
fixed it

karthik-tarento commented on src/main/java/com/igot/cb/service/PromotionalContentServiceImpl.java:
```diff
@@ -0,0 +1,349 @@
+            return response;
+        }
+        String errMsg = payloadValidation.validateAccessControlPayload(userGroupDetails);
+        if (org.apache.commons.lang.StringUtils.isNotBlank(errMsg)) {
```
imports needs to be optimized.

tarentomaheshvakkund replied:
done

karthik-tarento commented on src/main/java/com/igot/cb/service/PromotionalContentServiceImpl.java:
```diff
@@ -0,0 +1,349 @@
+     */
+    private void setFailedResponse(ApiResponse response, String errorMessage) {
+        setFailedResponse(response, errorMessage, HttpStatus.BAD_REQUEST);
+    }
```
these should be moved to util classes

tarentomaheshvakkund replied:
done

karthik-tarento commented on src/main/java/com/igot/cb/service/PromotionalContentServiceImpl.java:
```diff
@@ -0,0 +1,349 @@
+        response.getParams().setStatus(Constants.FAILED);
+        response.setResponseCode(httpStatus);
+        response.getParams().setErrMsg(errorMessage);
+    }
```
these should be moved to util classes

karthik-tarento commented on src/main/java/com/igot/cb/service/PromotionalContentServiceImpl.java:
```diff
@@ -0,0 +1,349 @@
+     * Modifies payload in-place and returns it.
+     */
+    @SuppressWarnings("unchecked")
+    private Map<String, Object> createUserGroupIds(Map<String, Object> payload) {
```
criteria list needs to be added into the user group and criteria key value pair needs to be added into the criteria list.. 
As of now, incoming payload will contain only the designation

karthik-tarento commented on src/main/java/com/igot/cb/service/PromotionalContentServiceImpl.java:
```diff
@@ -0,0 +1,349 @@
+            Map<String, Object> accessRuleData = new HashMap<>();
+            accessRuleData.put(Constants.CONTEXT_ID, contentId);
+            accessRuleData.put(Constants.CONTEXT_DATA, "");
+            accessRuleData.put(Constants.IS_ARCHIVED, false);
```
can we add this in the condtion of cq ?

tarentomaheshvakkund committed:
Code Review comments fixed

1. Caffeine cache is being used in the promotional content rule cache manager.
2. Pagination implemented for fetching the records from the cassandra tables- so that all the records and not fetched in one go.
3. Imports fixed.
4. Code format done.

Sahil-tarento commented on src/main/java/com/igot/cb/service/PromotionalContentServiceImpl.java:
```diff
@@ -0,0 +1,323 @@
+    @Override
+    public ApiResponse upsertPromotionalContentMetadata(Map<String, Object> userGroupDetails, String authToken) {
+        log.info("PromotionalContentServiceImpl::upsertPromotionalContentMetadata:inside");
+        ApiResponse response = ApiResponse.createDefaultResponse("api/promotionalcontent/metadata/upsert");
```
api/promotionalcontent/metadata/upsert
These should be came from constants

Sahil-tarento commented on src/main/java/com/igot/cb/service/PromotionalContentServiceImpl.java:
```diff
@@ -0,0 +1,323 @@
+    @Override
+    public ApiResponse getPromotionalContentForUsers(String authToken) {
+        log.info("PromotionalContentServiceImpl::getPromotionalContentForUsers:inside");
+        ApiResponse response = ApiResponse.createDefaultResponse("api/promotionalcontent/assignedto/users");
```
api/promotionalcontent/assignedto/users
Please move this to constants

Sahil-tarento commented on src/main/java/com/igot/cb/util/PayloadValidation.java:
```diff
@@ -17,6 +18,10 @@ public class PayloadValidation {
 
+    if (MapUtils.isEmpty(payload)) {
+         log.error("User group details are null or empty");
+         return "User group details cannot be null or empty";
```
Error msg should be came from constants file

tarentomaheshvakkund committed:
Added the hard coded string values to the constants file
